c/include/libsbp/cpp/message_handler.h

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright (C) 2019 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef SBP_CPP_MESSAGE_HANDLER_H_
+#define SBP_CPP_MESSAGE_HANDLER_H_
+
+#include <cassert>
+#include <array>
+
+#include <libsbp/cpp/state.h>
+#include <libsbp/cpp/message_traits.h>
+
+namespace sbp {
+
+/**
+ * A convenience type alias for class member functions that accept SBP message callbacks
+ */
+template<typename ClassT, typename ArgT>
+using CallbackMemFn = void (ClassT::*)(uint16_t, uint8_t, const ArgT &);
+
+/**
+ * A helper function for calling a C++ object member function from a libsbp callback.
+ *
+ * This function is registered with libsbp as a callback and calls a specific member function
+ * of an object from that callback with the SBP message decoded. The member function pointer
+ * is encoded into the template parameter, and the instance of the object is passed in via
+ * the `context` parameter.
+ *
+ * @tparam MsgT The message type to decode the payload as
+ * @tparam ClassT The class type to call the function on
```

Comment:
Is there a way to call the same function on multiple class types?  I find this is a common pattern for me (i.e do the same thing for any ephemeris message received)

Reply:
Not with how this PR currently stands. The problem with that all of the message types are distinct and there's little to no information relating the content of similar messages. The closest thing we could get is defining a base class that handles each of the distinct messages transforms them into a common type and then calls another member function that works on the common type.

Something like this:
```
class CommonEphemerisHandler : public sbp::MessageHandler<EphemerisMessage1, EphemerisMessage2> {
  public:
    struct CommonEphemerisType { ... };

    virtual void handle_common_ephemeris(const CommonEphemerisType& eph) = 0;

    void handle_sbp_msg(uint16_t sender_id, uint8_t message_length, const EphemerisMessage1& msg) {
      CommonEphemerisType eph = convertEph1(msg);
      handle_common_ephemeris(eph);
    }

    void handle_sbp_msg(uint16_t sender_id, uint8_t message_length, const EphemerisMessage2& msg) {
      CommonEphemerisType eph = convertEph2(msg);
      handle_common_ephemeris(eph);
    }
};

class CustomEphemerisHandler : public CommonEphemerisHandler {
  public:
    void handle_common_ephemeris(const CommonEphemerisType& eph) { ... }
};

class OtherEphemerisHandler : public CommonEphemerisHandler {
  public:
    void handle_common_ephemeris(const CommonEphemerisType& eph) { ... }
};
```